### ISSUE TITLE
[BUG] One of BlobContainerClient constructor overloads does not set clientSideEncryptionOptions

### ISSUE THREAD

**Issue by DeagleGross** [Azure/azure-sdk-for-net#44623]
### Library name and version

Azure.Storage.Blobs 12.20.0

### Describe the bug

[This constructor of BlobContainerClient](https://github.com/Azure/azure-sdk-for-net/blob/46a9641500d7f4a489125b0f7cd03c574832da2b/sdk/storage/Azure.Storage.Blobs/src/BlobContainerClient.cs#L187) does not set `_clientSideEncryptionOptions` on the client, and any operations will run without encryption specified.

### Expected behavior

client encryption options should be set

### Actual behavior

client encryption options are `null`

### Reproduction Steps

Run in debug
```
var client = new BlobContainerClient(
            connectionString: "UseDevelopmentStorage=true"
            blobContainerName: "enc-test",
            options: new SpecializedBlobClientOptions()
            {
                ClientSideEncryption = <some client side encryption options here>
            });
```
and make sure `client.ClientSideEncryption` is null (**should not be null**)

**Comment by nickliu-msft** [comment]
@DeagleGross Hi I was wondering, which operation are you trying to perform that requires client side encryption? 
Trying to understand if that operation will require encryption in the first place. Thanks

cc: @amnguye

**Comment by DeagleGross** [comment]
Hey @nickliu-msft , thanks for the fix.

Here is why we need this constructor API:
- we had such a code using legacy Azure Storage SDK where `BlobEncryptionPolicy` uses a manually constructed key:
```c#
string key = "primary key here";

var dataEncryptionKey = Convert.FromBase64String(key);
var blobEncryptionKey = new byte[64];
Array.Copy(dataEncryptionKey, blobEncryptionKey, Math.Min(blobEncryptionKey.Length, dataEncryptionKey.Length));

var blobRequestOptions = new BlobRequestOptions
{
    RetryPolicy = retryPolicy,
    EncryptionPolicy = new BlobEncryptionPolicy(new SymmetricKey("default", blobEncryptionKey), null)
};

// construct storage client here
```

- and here is how I migrated to new Azure Storage SDK:
```c#
public static ClientSideEncryptionOptions BuildEncryptionOptions(ISettingsProvider settingsProvider)
{
    var encryptionKey = GetEncryptionKey(settingsProvider);
    
    using var aes = Aes.Create(); // we can dispose AES, because `JsonWebKey` will only use the key of this object
    aes.Key = encryptionKey;
    
    var jsonWebKey = new JsonWebKey(aes);
    jsonWebKey.Id = "default";
    
    // we use protocol of version 1.0 to make it compatible with legacy SDK implementation (https://www.nuget.org/packages/Microsoft.Azure.Storage.Blob/). Otherwise we need to migrate blobs
    return new ClientSideEncryptionOptions(ClientSideEncryptionVersion.V1_0)
    {
        KeyEncryptionKey = new CryptographyClient(jsonWebKey, new LocalCryptographyClientOptions()), // only local cryptography enabled
        KeyWrapAlgorithm = "A256KW" // the same algorithm is used by legacy SDK (https://www.nuget.org/packages/Microsoft.Azure.Storage.Blob/)
    };
}

static byte[] GetEncryptionKey()
{
    var key = "<get a key here>";
    var dataEncryptionKey = Convert.FromBase64String(key);
    var blobEncryptionKey = new byte[32]; // AES can build with maximum 256-bit key, which is byte[32]
    Array.Copy(dataEncryptionKey, blobEncryptionKey, Math.Min(blobEncryptionKey.Length, dataEncryptionKey.Length));
    return blobEncryptionKey;
}
```

plus I needed to include **this reflection field assignment** to workaround current bug:
```c#
// sdk has an internal bug, where ClientSideEncryptionOptions is not set via this specific constructor overload
// https://github.com/Azure/azure-sdk-for-net/issues/44623
// therefore we need to set it via reflection. Please, check new versions and if bug is fixed - remove hacky-field assignment
if (encryptionOptions is not null)
{
    var clientSideEncryptionField = containerClient.GetType().GetField("_clientSideEncryption", BindingFlags.Instance | BindingFlags.NonPublic);
    clientSideEncryptionField!.SetValue(containerClient, encryptionOptions);
}
```

**Comment by nickliu-msft** [comment]
This PR fixes the above issue: [#45214](https://github.com/Azure/azure-sdk-for-net/pull/45214)
In the next release, this change will go out. Thanks

**Comment by DeagleGross** [comment]
thanks for a fast fix @nickliu-msft !
